File: python/y2019/day03.py
from aoc.core import solve, AocInput
from dataclasses import astuple, dataclass


@dataclass(frozen=True, slots=True)
class Position:
    irow: int
    icol: int

    def __iter__(self):
        return iter(astuple(self))

    @staticmethod
    def move(position: "Position", command: str) -> "Position":
        irow, icol = position
        match command:
            case "U":
                return Position(irow - 1, icol)
            case "D":
                return Position(irow + 1, icol)
            case "R":
                return Position(irow, icol + 1)
            case "L":
                return Position(irow, icol - 1)
            case _:
                position


def traceWire(path: str) -> dict[tuple[int, int], int]:
    positions = dict()
    irow, icol = Position(0, 0)
    distance = 0
    steps = path.split(",")
    for step in steps:
        for _ in range(int(step[1:])):
            irow, icol = Position.move(Position(irow, icol), step[0])
            distance += 1
            if positions.get((irow, icol)) is None:
                positions[(irow, icol)] = distance

    return positions


def part1(input: AocInput) -> int:
    wire1 = traceWire(input.lines[0])
    wire2 = traceWire(input.lines[1])
    return min(abs(p[0]) + abs(p[1]) for p in wire1.keys() if p in wire2)


def part2(input: AocInput) -> int:
    wire1 = traceWire(input.lines[0])
    wire2 = traceWire(input.lines[1])
    return min(wire1[p] + wire2[p] for p in wire1.keys() if p in wire2)


def main() -> None:
    YEAR = "y2019"
    DAY = "Day03"
    print(f"Running year {YEAR.replace('y', '')} day {DAY.replace('Day', '')}")
    solve(YEAR, DAY, part1)
    solve(YEAR, DAY, part2)

File: dotnet/2019/AdventOfCode.Y2019/Day03.cs
using AdventOfCode.Lib;

namespace AdventOfCode.Y2019;

[AocPuzzle(2019, 3, "Crossed Wires")]
public sealed class Day03 : IAocDay<int>
{
    public static int Part1(AocInput input)
    {
        var wire1 = Trace(input.AllLines[0]);
        var wire2 = Trace(input.AllLines[1]);
        return wire1.Keys.Where(p => wire2.ContainsKey(p)).Min(p => Math.Abs(p.Irow) + Math.Abs(p.Icol));
    }

    public static int Part2(AocInput input)
    {
        var wire1 = Trace(input.AllLines[0]);
        var wire2 = Trace(input.AllLines[1]);
        return wire1.Keys.Where(p => wire2.ContainsKey(p)).Min(p => wire1[p] + wire2[p]);
    }


    private static Dictionary<Position, int> Trace(string path)
    {
        Dictionary<Position, int> positions = [];
        var position = new Position(0, 0);
        var distance = 0;
        var steps = path.Split(',');
        foreach (var step in steps)
        {
            if (step is not [var command, .. var len])
            {
                continue;
            }

            for (var i = 0; i < int.Parse(len); i++)
            {
                position = Position.Move(position, command);
                distance++;
                if (positions.TryGetValue(position, out _) is false)
                {
                    positions[position] = distance;
                }
            }
        }

        return positions;
    }

    private readonly record struct Position(int Irow, int Icol)
    {
        public static Position Move(Position position, char command)
        {
            return command switch
            {
                'U' => position with { Irow = position.Irow - 1 },
                'D' => position with { Irow = position.Irow + 1 },
                'R' => position with { Icol = position.Icol + 1 },
                'L' => position with { Icol = position.Icol - 1 },
                _ => throw new ArgumentException("Unknown Command")
            };
        }
    }
}